### PR TITLE
Create new Bootstrapper API

### PR DESCRIPTION
**Pull request by brianlagunas**
﻿## Description of Change

implemented a new bootstrapper API to replace the old legacy bootstrapper.

### Bugs Fixed

- #2061

### API Changes

Added:
- PrismBootstrapperBase
- PrismBootstrapper (container specific)

### PR Checklist

- [ ] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

File: e2e/Wpf/HelloWorld.Bootstraper/App.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    /// Interaction logic for App.xaml
+    /// </summary>
+    public partial class App : Application
+    {
+        protected override void OnStartup(StartupEventArgs e)
+        {
+            base.OnStartup(e);
+
+            var bs = new Bootstrapper();
+            bs.Run();
```
Comment (dansiegel) [inline e2e/Wpf/HelloWorld.Bootstraper/App.xaml.cs:15]: you're right it is bs... haha

File: e2e/Wpf/HelloWorld.Bootstraper/HelloWorld.Bootstrapper.csproj [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <DefaultItemExcludes>..\HelloWorld\bin\**;..\HelloWorld\obj\**;..\HelloWorld\App.*</DefaultItemExcludes>
+  </PropertyGroup>
+  
+  <ItemGroup>
+    <Page Include="..\HelloWorld\**\*.xaml" LinkBase="%(RecurseDir)" Generator="MSBuild:Compile" Exclude="$(DefaultItemExcludes)" />
+    <Compile Include="..\HelloWorld\**\*.cs" LinkBase="%(RecurseDir)" Exclude="$(DefaultItemExcludes)" />
+  </ItemGroup>
```
Comment (dansiegel) [inline e2e/Wpf/HelloWorld.Bootstraper/HelloWorld.Bootstrapper.csproj:17]: happy you're learning the power of MSBuild and SDK Style projects!

File: e2e/Wpf/HelloWorld.Bootstraper/App.xaml.cs (hunk 1 above)
Comment (brianlagunas) [reply e2e/Wpf/HelloWorld.Bootstraper/App.xaml.cs:15]: that's my joke every time I write this line of code :)

File: src/Wpf/Prism.DryIoc.Wpf/PrismBootstrapper.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+{
+    public abstract class PrismBootstrapper : PrismBootstrapperBase
+    {
+        /// <summary>
+        /// Create <see cref="Rules" /> to alter behavior of <see cref="IContainer" />
+        /// </summary>
+        /// <returns>An instance of <see cref="Rules" /></returns>
+        protected virtual Rules CreateContainerRules() => Rules.Default.WithAutoConcreteTypeResolution()
+                                                                       .With(Made.Of(FactoryMethod.ConstructorWithResolvableArguments))
+                                                                       .WithDefaultIfAlreadyRegistered(IfAlreadyRegistered.Replace);
```
Comment (dansiegel) [inline src/Wpf/Prism.DryIoc.Wpf/PrismBootstrapper.cs:15]: we're doing this in too many places... I'm wondering if we should move the default Rules to a static property off of the DryIocContainerExtension? Granted we would probably be doing some #if SomePlatform to to handle minor differences between mobile/desktop

File: src/Wpf/Prism.Wpf/PrismBootstrapperBase.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        /// </summary>
+        public IContainerProvider Container => _containerExtension;
+
+        /// <summary>
+        /// Gets the shell user interface
+        /// </summary>
+        /// <value>The shell user interface.</value>
+        protected Window Shell { get; set; }
+
+
```
Comment (dansiegel) [inline src/Wpf/Prism.Wpf/PrismBootstrapperBase.cs:32]: should really be a single line of empty space

File: src/Wpf/Prism.Wpf/PrismBootstrapperBase.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        /// <summary>
+        /// Run the initialization process.
+        /// </summary>
+        void InitializeInternal()
+        {
+            ConfigureViewModelLocator();
+            Initialize();
+            OnInitialized();
+        }
```
Comment (dansiegel) [inline src/Wpf/Prism.Wpf/PrismBootstrapperBase.cs:46]: this seems silly to me... really InitializeInternal is the Run... don't really see a point to having a public Run that is only calling an InitializeInternal

File: src/Wpf/Prism.Wpf/PrismBootstrapperBase.cs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        /// May be overwritten in a derived class to add specific mappings required by the application.
+        /// </summary>
+        /// <returns>The <see cref="RegionAdapterMappings"/> instance containing all the mappings.</returns>
+        protected virtual void ConfigureRegionAdapterMappings(RegionAdapterMappings regionAdapterMappings)
+        {
+            if (regionAdapterMappings != null)
+            {
+                regionAdapterMappings.RegisterMapping(typeof(Selector), _containerExtension.Resolve<SelectorRegionAdapter>());
+                regionAdapterMappings.RegisterMapping(typeof(ItemsControl), _containerExtension.Resolve<ItemsControlRegionAdapter>());
+                regionAdapterMappings.RegisterMapping(typeof(ContentControl), _containerExtension.Resolve<ContentControlRegionAdapter>());
```
Comment (dansiegel) [inline src/Wpf/Prism.Wpf/PrismBootstrapperBase.cs:168]: this should be updated...

```cs
regionAdapterMappings.RegisterMapping<Selector, SelectorRegionAdapter>();
```

File: src/Wpf/Prism.DryIoc.Wpf/PrismBootstrapper.cs (hunk 3 above)
Comment (brianlagunas) [reply src/Wpf/Prism.DryIoc.Wpf/PrismBootstrapper.cs:15]: yeah, that should probably be a separate issue. We have the same problem for other things like region adapters, require types, etc. that are being used on both the App and Bootstrapper classes.

File: src/Wpf/Prism.Wpf/PrismBootstrapperBase.cs (hunk 5 above)
Comment (brianlagunas) [reply src/Wpf/Prism.Wpf/PrismBootstrapperBase.cs:46]: Really just for consistency between the two approaches.

**Pull request merged**